graylog2-server/src/main/java/org/graylog/plugins/views/search/validation/TokenCollectingQueryParser.java

```diff
@@ -31,4 +44,74 @@ public TokenCollectingQueryParser(String defaultFieldName, Analyzer analyzer) {
     public List<ImmutableToken> getTokens() {
         return ((CollectingQueryParserTokenManager) super.token_source).getTokens();
     }
+
+    public Map<Query, Collection<ImmutableToken>> getTokenLookup() {
+        return tokenLookup;
+    }
+
+    @Override
+    protected Query newFieldQuery(Analyzer analyzer, String field, String queryText, boolean quoted) throws ParseException {
+        return saveQueryLookupTokens(super.newFieldQuery(analyzer, field, queryText, quoted));
+    }
+
+    @Override
+    protected Query newPrefixQuery(Term prefix) {
+        return saveQueryLookupTokens(super.newPrefixQuery(prefix));
+    }
+
+    @Override
+    protected Query newRegexpQuery(Term regexp) {
+        return saveQueryLookupTokens(super.newRegexpQuery(regexp));
+    }
+
+    @Override
+    protected Query newFuzzyQuery(Term term, float minimumSimilarity, int prefixLength) {
+        return saveQueryLookupTokens(super.newFuzzyQuery(term, minimumSimilarity, prefixLength));
+    }
+
+    @Override
+    protected Query newMatchAllDocsQuery() {
+        return saveQueryLookupTokens(super.newMatchAllDocsQuery());
+    }
+
+    @Override
+    protected Query newWildcardQuery(Term t) {
+        return saveQueryLookupTokens(super.newWildcardQuery(t));
+    }
+
+    @Override
+    protected Query newSynonymQuery(TermAndBoost[] terms) {
+        return saveQueryLookupTokens(super.newSynonymQuery(terms));
+    }
+
+    @Override
+    protected Query newGraphSynonymQuery(Iterator<Query> queries) {
+        return saveQueryLookupTokens(super.newGraphSynonymQuery(queries));
+    }
+
+    @Override
+    protected Query newTermQuery(Term term, float boost) {
+        return saveQueryLookupTokens(super.newTermQuery(term, boost));
+    }
+
+    @Override
+    protected Query newRangeQuery(String field, String part1, String part2, boolean startInclusive, boolean endInclusive) {
+        return saveQueryLookupTokens(super.newRangeQuery(field, part1, part2, startInclusive, endInclusive));
+    }
+
+    /**
+     * This method persists all newly discovered tokens that may be referenced in the query to a lookup for
+     * later processing. This is the only place that binds query and its tokens together.
+     *
+     * This would be also good place to collect all subqueries if needed for any feature.
+     */
+    private Query saveQueryLookupTokens(Query query) {
+        if (!tokenLookup.containsKey(query)) {
```

Comment:
I am just wondering about identity of the queries in this place...
If we have two queries with the same query string, and they are equal according to equals() method, should we really avoid processing the second one?
You know the details better, but there is a chance that IdentityHashMap would be a better implementation of tokenLookup...

Reply:
Good catch! Had to adapt the logic a bit and extend the lucene query parser so it correctly handles identity of queries in boolean queries, but it all seems to work now. Thanks!